Error due to missing libgmp when installing nokogiri gem on ruby:2.5-slim

Moving [from discussion](https://github.com/docker-library/ruby/commit/eae22dc2dfe30b48f1633912396110539e1b8d49#commitcomment-32559528) on eae22dc2dfe30b48f1633912396110539e1b8d49. It seems that libgmp is somehow missing on the latest `ruby:2.5-slim` image. Others reported similar behavior on the alpine image.

```
$ docker pull ruby:2.5-slim
2.5-slim: Pulling from library/ruby
Digest: sha256:c8d66ac4a1fa252df07f505f9dc02b5ee203f33b66406bfddca5c55c1d5f4b58
Status: Image is up to date for ruby:2.5-slim

$ docker run -it --rm ruby:2.5-slim /bin/bash
root@6465f040df18:/# apt-get update >/dev/null 2>&1
root@6465f040df18:/# apt-get -y install -qq -o=Dpkg::Use-Pty=0 build-essential libxslt-dev libxml2-dev >/dev/null 2>&1
root@6465f040df18:/# gem install nokogiri
Fetching nokogiri-1.10.1.gem
Fetching mini_portile2-2.4.0.gem
Successfully installed mini_portile2-2.4.0
Building native extensions. This could take a while...
ERROR:  Error installing nokogiri:
        ERROR: Failed to build gem native extension.

    current directory: /usr/local/bundle/gems/nokogiri-1.10.1/ext/nokogiri
/usr/local/bin/ruby -I /usr/local/lib/ruby/site_ruby/2.5.0 -r ./siteconf20190303-931-p72cjl.rb extconf.rb
checking if the C compiler accepts ... *** extconf.rb failed ***
Could not create Makefile due to some reason, probably lack of necessary
libraries and/or headers.  Check the mkmf.log file for more details.  You may
need configuration options.

Provided configuration options:
        --with-opt-dir
        --without-opt-dir
        --with-opt-include
        --without-opt-include=${opt-dir}/include
        --with-opt-lib
        --without-opt-lib=${opt-dir}/lib
        --with-make-prog
        --without-make-prog
        --srcdir=.
        --curdir
        --ruby=/usr/local/bin/$(RUBY_BASE_NAME)
        --help
        --clean
/usr/local/lib/ruby/2.5.0/mkmf.rb:456:in `try_do': The compiler failed to generate an executable file. (RuntimeError)
You have to install development tools first.
        from /usr/local/lib/ruby/2.5.0/mkmf.rb:574:in `block in try_compile'
        from /usr/local/lib/ruby/2.5.0/mkmf.rb:521:in `with_werror'
        from /usr/local/lib/ruby/2.5.0/mkmf.rb:574:in `try_compile'
        from extconf.rb:138:in `nokogiri_try_compile'
        from extconf.rb:162:in `block in add_cflags'
        from /usr/local/lib/ruby/2.5.0/mkmf.rb:632:in `with_cflags'
        from extconf.rb:161:in `add_cflags'
        from extconf.rb:416:in `<main>'

To see why this extension failed to compile, please check the mkmf.log which can be found here:

  /usr/local/bundle/extensions/x86_64-linux/2.5.0/nokogiri-1.10.1/mkmf.log

extconf failed, exit code 1

Gem files will remain installed in /usr/local/bundle/gems/nokogiri-1.10.1 for inspection.
Results logged to /usr/local/bundle/extensions/x86_64-linux/2.5.0/nokogiri-1.10.1/gem_make.out


root@6465f040df18:/# cat /usr/local/bundle/extensions/x86_64-linux/2.5.0/nokogiri-1.10.1/mkmf.log
"gcc -o conftest -I/usr/local/include/ruby-2.5.0/x86_64-linux -I/usr/local/include/ruby-2.5.0/ruby/backward -I/usr/local/include/ruby-2.5.0 -I.    -O3 -ggdb3 -Wall -Wextra -Wno-unused-parameter -Wno-parentheses -Wno-long-long -Wno-missing-field-initializers -Wno-tautological-compare -Wno-parentheses-equality -Wno-constant-logical-operand -Wno-self-assign -Wunused-variable -Wimplicit-int -Wpointer-arith -Wwrite-strings -Wdeclaration-after-statement -Wimplicit-function-declaration -Wdeprecated-declarations -Wmisleading-indentation -Wno-packed-bitfield-compat -Wsuggest-attribute=noreturn -Wsuggest-attribute=format -Wduplicated-cond -Wno-maybe-uninitialized  -fPIC  conftest.c  -L. -L/usr/local/lib -Wl,-rpath,/usr/local/lib -L. -fstack-protector -rdynamic -Wl,-export-dynamic     -Wl,-rpath,/usr/local/lib -L/usr/local/lib -lruby  -lpthread -lgmp -ldl -lcrypt -lm   -lc "
/usr/bin/ld: cannot find -lgmp
collect2: error: ld returned 1 exit status
checked program was:
/* begin */
1: #include "ruby.h"
2: 
3: int main(int argc, char **argv)
4: {
5:   return 0;
6: }
/* end */
```

It seems that `libgmp-dev` somehow didn't make it into the image, because I can manually install it and then things work okay:

```
root@6465f040df18:/# apt-get install libgmp-dev
Reading package lists... Done
Building dependency tree       
Reading state information... Done
The following additional packages will be installed:
  libgmpxx4ldbl
Suggested packages:
  gmp-doc libgmp10-doc libmpfr-dev
The following NEW packages will be installed:
  libgmp-dev libgmpxx4ldbl
0 upgraded, 2 newly installed, 0 to remove and 10 not upgraded.
Need to get 653 kB of archives.
After this operation, 1971 kB of additional disk space will be used.
Do you want to continue? [Y/n] Y
Get:1 http://deb.debian.org/debian stretch/main amd64 libgmpxx4ldbl amd64 2:6.1.2+dfsg-1 [22.2 kB]
Get:2 http://deb.debian.org/debian stretch/main amd64 libgmp-dev amd64 2:6.1.2+dfsg-1 [631 kB]
Fetched 653 kB in 0s (2493 kB/s)    
debconf: delaying package configuration, since apt-utils is not installed
Selecting previously unselected package libgmpxx4ldbl:amd64.
(Reading database ... 13068 files and directories currently installed.)
Preparing to unpack .../libgmpxx4ldbl_2%3a6.1.2+dfsg-1_amd64.deb ...
Unpacking libgmpxx4ldbl:amd64 (2:6.1.2+dfsg-1) ...
Selecting previously unselected package libgmp-dev:amd64.
Preparing to unpack .../libgmp-dev_2%3a6.1.2+dfsg-1_amd64.deb ...
Unpacking libgmp-dev:amd64 (2:6.1.2+dfsg-1) ...
Processing triggers for libc-bin (2.24-11+deb9u3) ...
Setting up libgmpxx4ldbl:amd64 (2:6.1.2+dfsg-1) ...
Setting up libgmp-dev:amd64 (2:6.1.2+dfsg-1) ...
Processing triggers for libc-bin (2.24-11+deb9u3) ...

root@6465f040df18:/# gem install nokogiri
Building native extensions. This could take a while...
Successfully installed nokogiri-1.10.1
1 gem installed
```

Alternatively, installing `libgmp3-dev` also works to fix the issue.

--- a/2.3/alpine3.7/Dockerfile
+++ b/2.3/alpine3.7/Dockerfile
@@ -1,5 +1,8 @@
 FROM alpine:3.7
 
+RUN apk add --no-cache \
+		gmp-dev
+
 # skip installing gem documentation
 RUN mkdir -p /usr/local/etc \
 	&& { \
@@ -28,7 +31,6 @@ RUN set -ex \
 		gcc \
 		gdbm-dev \
 		glib-dev \
-		gmp-dev \
 		libc-dev \
 		libffi-dev \
 		libxml2-dev \

--- a/2.3/alpine3.8/Dockerfile
+++ b/2.3/alpine3.8/Dockerfile
@@ -1,5 +1,8 @@
 FROM alpine:3.8
 
+RUN apk add --no-cache \
+		gmp-dev
+
 # skip installing gem documentation
 RUN mkdir -p /usr/local/etc \
 	&& { \
@@ -28,7 +31,6 @@ RUN set -ex \
 		gcc \
 		gdbm-dev \
 		glib-dev \
-		gmp-dev \
 		libc-dev \
 		libffi-dev \
 		libxml2-dev \

--- a/2.3/jessie/slim/Dockerfile
+++ b/2.3/jessie/slim/Dockerfile
@@ -6,6 +6,7 @@ RUN apt-get update \
 		ca-certificates \
 		libffi-dev \
 		libgdbm3 \
+		libgmp-dev \
 		libssl-dev \
 		libyaml-dev \
 		procps \
@@ -37,7 +38,6 @@ RUN set -ex \
 		libbz2-dev \
 		libgdbm-dev \
 		libglib2.0-dev \
-		libgmp-dev \
 		libncurses-dev \
 		libreadline-dev \
 		libxml2-dev \

--- a/2.3/stretch/slim/Dockerfile
+++ b/2.3/stretch/slim/Dockerfile
@@ -6,6 +6,7 @@ RUN apt-get update \
 		ca-certificates \
 		libffi-dev \
 		libgdbm3 \
+		libgmp-dev \
 		libssl1.0-dev \
 		libyaml-dev \
 		procps \
@@ -37,7 +38,6 @@ RUN set -ex \
 		libbz2-dev \
 		libgdbm-dev \
 		libglib2.0-dev \
-		libgmp-dev \
 		libncurses-dev \
 		libreadline-dev \
 		libxml2-dev \

--- a/2.4/alpine3.8/Dockerfile
+++ b/2.4/alpine3.8/Dockerfile
@@ -1,5 +1,8 @@
 FROM alpine:3.8
 
+RUN apk add --no-cache \
+		gmp-dev
+
 # skip installing gem documentation
 RUN mkdir -p /usr/local/etc \
 	&& { \
@@ -28,7 +31,6 @@ RUN set -ex \
 		gcc \
 		gdbm-dev \
 		glib-dev \
-		gmp-dev \
 		libc-dev \
 		libffi-dev \
 		libxml2-dev \

--- a/2.4/alpine3.9/Dockerfile
+++ b/2.4/alpine3.9/Dockerfile
@@ -1,5 +1,8 @@
 FROM alpine:3.9
 
+RUN apk add --no-cache \
+		gmp-dev
+
 # skip installing gem documentation
 RUN mkdir -p /usr/local/etc \
 	&& { \
@@ -28,7 +31,6 @@ RUN set -ex \
 		gcc \
 		gdbm-dev \
 		glib-dev \
-		gmp-dev \
 		libc-dev \
 		libffi-dev \
 		libxml2-dev \

--- a/2.4/jessie/slim/Dockerfile
+++ b/2.4/jessie/slim/Dockerfile
@@ -6,6 +6,7 @@ RUN apt-get update \
 		ca-certificates \
 		libffi-dev \
 		libgdbm3 \
+		libgmp-dev \
 		libssl-dev \
 		libyaml-dev \
 		procps \
@@ -37,7 +38,6 @@ RUN set -ex \
 		libbz2-dev \
 		libgdbm-dev \
 		libglib2.0-dev \
-		libgmp-dev \
 		libncurses-dev \
 		libreadline-dev \
 		libxml2-dev \

--- a/2.4/stretch/slim/Dockerfile
+++ b/2.4/stretch/slim/Dockerfile
@@ -6,6 +6,7 @@ RUN apt-get update \
 		ca-certificates \
 		libffi-dev \
 		libgdbm3 \
+		libgmp-dev \
 		libssl-dev \
 		libyaml-dev \
 		procps \
@@ -37,7 +38,6 @@ RUN set -ex \
 		libbz2-dev \
 		libgdbm-dev \
 		libglib2.0-dev \
-		libgmp-dev \
 		libncurses-dev \
 		libreadline-dev \
 		libxml2-dev \

--- a/2.5/alpine3.8/Dockerfile
+++ b/2.5/alpine3.8/Dockerfile
@@ -1,5 +1,8 @@
 FROM alpine:3.8
 
+RUN apk add --no-cache \
+		gmp-dev
+
 # skip installing gem documentation
 RUN mkdir -p /usr/local/etc \
 	&& { \
@@ -28,7 +31,6 @@ RUN set -ex \
 		gcc \
 		gdbm-dev \
 		glib-dev \
-		gmp-dev \
 		libc-dev \
 		libffi-dev \
 		libxml2-dev \

--- a/2.5/alpine3.9/Dockerfile
+++ b/2.5/alpine3.9/Dockerfile
@@ -1,5 +1,8 @@
 FROM alpine:3.9
 
+RUN apk add --no-cache \
+		gmp-dev
+
 # skip installing gem documentation
 RUN mkdir -p /usr/local/etc \
 	&& { \
@@ -28,7 +31,6 @@ RUN set -ex \
 		gcc \
 		gdbm-dev \
 		glib-dev \
-		gmp-dev \
 		libc-dev \
 		libffi-dev \
 		libxml2-dev \

--- a/2.5/stretch/slim/Dockerfile
+++ b/2.5/stretch/slim/Dockerfile
@@ -6,6 +6,7 @@ RUN apt-get update \
 		ca-certificates \
 		libffi-dev \
 		libgdbm3 \
+		libgmp-dev \
 		libssl-dev \
 		libyaml-dev \
 		procps \
@@ -37,7 +38,6 @@ RUN set -ex \
 		libbz2-dev \
 		libgdbm-dev \
 		libglib2.0-dev \
-		libgmp-dev \
 		libncurses-dev \
 		libreadline-dev \
 		libxml2-dev \

--- a/2.6/alpine3.8/Dockerfile
+++ b/2.6/alpine3.8/Dockerfile
@@ -1,5 +1,8 @@
 FROM alpine:3.8
 
+RUN apk add --no-cache \
+		gmp-dev
+
 # skip installing gem documentation
 RUN mkdir -p /usr/local/etc \
 	&& { \
@@ -27,7 +30,6 @@ RUN set -ex \
 		gcc \
 		gdbm-dev \
 		glib-dev \
-		gmp-dev \
 		libc-dev \
 		libffi-dev \
 		libxml2-dev \

--- a/2.6/alpine3.9/Dockerfile
+++ b/2.6/alpine3.9/Dockerfile
@@ -1,5 +1,8 @@
 FROM alpine:3.9
 
+RUN apk add --no-cache \
+		gmp-dev
+
 # skip installing gem documentation
 RUN mkdir -p /usr/local/etc \
 	&& { \
@@ -27,7 +30,6 @@ RUN set -ex \
 		gcc \
 		gdbm-dev \
 		glib-dev \
-		gmp-dev \
 		libc-dev \
 		libffi-dev \
 		libxml2-dev \

--- a/2.6/stretch/slim/Dockerfile
+++ b/2.6/stretch/slim/Dockerfile
@@ -6,6 +6,7 @@ RUN apt-get update \
 		ca-certificates \
 		libffi-dev \
 		libgdbm3 \
+		libgmp-dev \
 		libssl-dev \
 		libyaml-dev \
 		procps \
@@ -36,7 +37,6 @@ RUN set -ex \
 		libbz2-dev \
 		libgdbm-dev \
 		libglib2.0-dev \
-		libgmp-dev \
 		libncurses-dev \
 		libreadline-dev \
 		libxml2-dev \

--- a/Dockerfile-alpine.template
+++ b/Dockerfile-alpine.template
@@ -1,5 +1,8 @@
 FROM alpine:%%PLACEHOLDER%%
 
+RUN apk add --no-cache \
+		gmp-dev
+
 # skip installing gem documentation
 RUN mkdir -p /usr/local/etc \
 	&& { \
@@ -28,7 +31,6 @@ RUN set -ex \
 		gcc \
 		gdbm-dev \
 		glib-dev \
-		gmp-dev \
 		libc-dev \
 		libffi-dev \
 		libxml2-dev \

--- a/Dockerfile-slim.template
+++ b/Dockerfile-slim.template
@@ -6,6 +6,7 @@ RUN apt-get update \
 		ca-certificates \
 		libffi-dev \
 		libgdbm3 \
+		libgmp-dev \
 		libssl-dev \
 		libyaml-dev \
 		procps \
@@ -37,7 +38,6 @@ RUN set -ex \
 		libbz2-dev \
 		libgdbm-dev \
 		libglib2.0-dev \
-		libgmp-dev \
 		libncurses-dev \
 		libreadline-dev \
 		libxml2-dev \

